File: elixir/anagram/lib/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    base = String.downcase(base)
    base_letters = letters(base)

    candidates
    |> Enum.filter(fn w ->
      w = String.downcase(w)
      w != base && letters(w) == base_letters
    end)

  end

  defp letters(s), do: s |> String.graphemes |> Enum.sort
end

File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.Char  (toLower)
import Data.List  (sort)

anagramsFor :: String -> [String] -> [String]
anagramsFor xs xss = anagrams
  where xs' = map toLower xs
        candidates = filter (\s -> map toLower s /= xs') xss
        sortedXs = sort xs'
        anagrams = filter (\s ->  sortedXs == (sort $ map toLower s)) candidates
